### ISSUE TITLE
XrdTpc PULL checksum issues

### ISSUE THREAD

**Issue by esindril** [xrootd/xrootd#1404]
Are there any known issues with TPC PULL where the final checksum does not match the user provided one?
We are running EOS on top of XRootD 4.12.7 and LHCb started using HTTP TPC in production but some of the TPC PULL transfers fail eventually since the final checksum does not match the user provided one. FTS later retries using probably TPC push and the transfers succeeds with the correct checksum.

In the EOS logs we don't see any errors whatsoever. For example on the disk node when such a file with the wrong checksum is written, we get a clean and successful transfer log:

```
210219 05:10:34 26159 TPC_PullRequest: event=PULL_START, local=/eos/lhcb/grid/prod/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122396_00001840_1.rxchad.strip.dst, remote=https://lhcbwebdav-kit.gridka.de:2880/pnfs/gridka.de/lhcb/LHCb-Disk/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122
396_00001840_1.rxchad.strip.dst, user=(anonymous); Starting a push request
210219 05:10:34 time=1613707834.930317 func=open                     level=INFO  logid=6a336346-7268-11eb-a7ac-a4bf0114cb20 unit=fst@st-048-bbead64b.cern.ch:1095 tid=00007f80735fb700 source=XrdFstOfsFile:120              tident=? sec=      uid=0 gid=0 name= geo="" path=/eos/lhcb/grid/prod/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122396_00001840_1.rxchad.strip.dst info=cap.sym=<...>&cap.msg=<...>&mgm.logid=6a336346-7268-11eb-a7ac-a4bf0114cb20&mgm.replicaindex=0&mgm.replicahead=0&mgm.etag="342790846266998784:00000000"&mgm.id=4c1d6756&authz=<...> open_mode=201
210219 05:10:34 time=1613707834.930530 func=ProcessCapOpaque         level=INFO  logid=6a336346-7268-11eb-a7ac-a4bf0114cb20 unit=fst@st-048-bbead64b.cern.ch:1095 tid=00007f80735fb700 source=XrdFstOfsFile:2253             tident=? sec=(null) uid=99 gid=99 name=(null) geo="" capability=&mgm.access=create&mgm.ruid=7947&mgm.rgid=1470&mgm.uid=99&mgm.gid=99&mgm.path=/eos/lhcb/grid/prod/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122396_00001840_1.rxchad.strip.dst&mgm.manager=eoslhcb-qdb-52bedb7c30.cern.ch:1094&mgm.fid=4c1d6756&mgm.cid=113766179&mgm.sec=https|lhcbprod|[2001:1458:301:cd::100:1ab]||||/DC=ch/DC=cern/OU=Organic Units/OU=Users/CN=fstagni/CN=693025/CN=Federico Stagni|&mgm.lid=1048850&mgm.bookingsize=1000000000&mgm.fsid=11575&mgm.url0=root://st-048-bbead64b.cern.ch:1095//&mgm.fsid0=11575&mgm.url1=root://p06636710f33060.cern.ch:1095//&mgm.fsid1=10611&cap.valid=1613711434
210219 05:10:34 time=1613707834.930577 func=open                     level=INFO  logid=6a336346-7268-11eb-a7ac-a4bf0114cb20 unit=fst@st-048-bbead64b.cern.ch:1095 tid=00007f80735fb700 source=XrdFstOfsFile:198              tident=? sec=(null) uid=7947 gid=1470 name=nobody geo="" ns_path=/eos/lhcb/grid/prod/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122396_00001840_1.rxchad.strip.dst fst_path=/data05/0001f2d3/4c1d6756
210219 05:10:34 time=1613707834.930926 func=open                     level=INFO  logid=6a336346-7268-11eb-a7ac-a4bf0114cb20 unit=fst@st-048-bbead64b.cern.ch:1095 tid=00007f80735fb700 source=XrdFstOfsFile:468              tident=? sec=(null) uid=7947 gid=1470 name=nobody geo="" fst_path=/data05/0001f2d3/4c1d6756 open-mode=301 create-mode=41a4 layout-name=replica oss-opaque=&mgm.lid=1048850&mgm.bookingsize=1000000000
210219 05:10:34 time=1613707834.930943 func=Open                     level=INFO  logid=6a336346-7268-11eb-a7ac-a4bf0114cb20 unit=fst@st-048-bbead64b.cern.ch:1095 tid=00007f80735fb700 source=ReplicaParLayout:104           tident=? sec=      uid=0 gid=0 name= geo="" replica_head=0, replica_index=0
....
210219 05:10:41 time=1613707841.983478 func=VerifyChecksum           level=INFO  logid=6a336346-7268-11eb-a7ac-a4bf0114cb20 unit=fst@st-048-bbead64b.cern.ch:1095 tid=00007f80735fb700 source=XrdFstOfsFile:3017             tident=? sec=      uid=7947 gid=1470 name=nobody geo="" (write) checksum type: adler checksum hex: 1edadd8c requested-checksum hex: -none-
210219 05:10:42 time=1613707842.071020 func=_close                   level=INFO  logid=6a336346-7268-11eb-a7ac-a4bf0114cb20 unit=fst@st-048-bbead64b.cern.ch:1095 tid=00007f80735fb700 source=XrdFstOfsFile:1807             tident=? sec=      uid=7947 gid=1470 name=nobody geo="" msg="done close" rc=0 errc=0
210219 05:10:42 26159 TPC_PullRequest: event=TRANSFER_SUCCESS, local=/eos/lhcb/grid/prod/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122396_00001840_1.rxchad.strip.dst, remote=https://lhcbwebdav-kit.gridka.de:2880/pnfs/gridka.de/lhcb/LHCb-Disk/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122396_00001840_1.rxchad.strip.dst, user=(anonymous), bytes_transferred=180232171, tpc_status=200
```

And the corresponding error in FTS reads:
```
INFO Fri, 19 Feb 2021 05:10:42 +0100; [1613707842071] BOTH http_plugin TRANSFER:EXIT https://lhcbwebdav-kit.gridka.de:2880/pnfs/gridka.de/lhcb/LHCb-Disk/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122396_00001840_1.rxchad.strip.dst => https://eoslhcb.cern.ch/eos/lhcb/grid/prod/lhcb/MC/2012/RXCHAD.STRIP.DST/00122396/0000/00122396_00001840_1.rxchad.strip.dst
ERR Fri, 19 Feb 2021 05:10:42 +0100; Non recoverable error: [5] DESTINATION CHECKSUM MISMATCH User-defined and destination ADLER32 do not match (fc21471b != 1edadd8c)
```

The EOS OFS layer is quite robust when it comes such write operations, so I would suspect some possible issues in the XrdTpc layer related to curl reading from remote destination/ buffering etc. Does this ring a bell?

Thanks!

**Comment by snafus** [comment]
Dear @esindril ,
  I'm not sure if this is related (and have not fully finished trying to diagnose the problem), but at RAL I am seeing checksum issues in our Test TPC gateway (5.1.0-rc7)
Our issue, is also with PULL mode.
It appears that sometimes, at 1MiB points in the file, the preceding n*1024 bytes from the ith MiB are copied into the ith MiB position. It looks like i == n in most cases, but I'm still working through the logic ... 
It is not quite so straightforward, as we get different checksums from repeated tests, but they quite often match each other - i.e. the corruption is happening in similar, but slightly different ways, but apparently on these 1MiB boundaries.

I note these are in async io transfers (I don't recall if we were always doing this for RAL's TPC tests.
Sorry if it's not relevant, and only a partial diagnosis,  but I hope it's useful to add a similar datapoint here.

Cheers,
James

**Comment by abh3** [comment]
I would dare say this is due to errors in the buffering logic in the TPC 
layer. That layer tries to accumulate 1MB of data before sending it off to 
the filesystem, ostensibly because anything less may cause issues with 
Ceph (I don't know why this is the case but the comments say it is). 
That's a tricky affair and when I looked at the code I was mystified by 
it's logic. Given the complexity here I would say that is where the 
problem occurs and would only affect TPC Pull requests.

This line of thought is further bolstered by corresponding to what you 
have observed.

Andy

**Comment by esindril** [comment]
Thanks to Mihai from the FTS team, I managed to catch one such corrupted file and compared its contents with the correct one. I seems the beginning of every MB is corrupted, for example bytes that differ in this particular case are:
1MB -> 1MB + 5952
2MB -> 2MB + 6688
3MB -> 3MB + 5728
....

403MB -> 403MB + 15632

I have the full info if this is necessary. The full size of the file is: 423436359 bytes
Actual logs from the FST transfer: 
https://fts-pilot-010.cern.ch:8449/var/log/fts3/transfers/2021-02-23/webdav.lcg.triumf.ca__eosatlas.cern.ch/2021-02-23-0948__webdav.lcg.triumf.ca__eosatlas.cern.ch__1448324283__9c8bd9bc-75bb-11eb-bc6b-fa163e14af38

In the meantime, similar errors were identified also in EOSATLAS with apparently a higher incidence rate once we moved to XRootD 4.12.7.

**Comment by xrootd-dev** [comment]
Indeed, I have reached out to the developer indicating that there is a 
problem in the buffer coalescing code in XrdTpc. I am waiting for a 
response.

Andy

**Comment by abh3** [comment]
BTY, the code has substantially changed in R5.1.0 so this problem may have 
been fixed relative to you. I don't think this fix can be backported as 
it's pretty extensive. Anyway, let's see what the developer says.

**Comment by esindril** [comment]
Hi Andy,

We'll move very soon to XrootD 5.1.0 so if this is already fixed even better. But for the moment we don't have any production instance running with XRootD 5 so I can not draw any comparison. Moreover, this does not seem to be a systemic error but triggered only in certain conditions.

Thanks!

**Comment by snafus** [comment]
Hi @abh3,
  Thanks for pointing me to 5.1.0. Unfortunately, at RAL, it looks to be reproducible in 5.1.0 as well. 
  with 'x' number of btyes at each MiB boundary overwritten by prior bytes. 
  e.g:
  Start bad, -> + btyes
1048576 -> + 2056
2097152 -> + 2056
3145728 -> + 2056
...

If I can find anything further that might be useful from 5.1.0 I'll try to update here.

Cheers,
James

**Comment by bbockelm** [comment]
@snafus @esindril -

So, the common criteria are:

1.  When it "overwrites", it does this starting on 1MB.  That is, it's likely not a race condition that develops in the middle of the file.
2.  When it overwrites, it does this consistently on every buffer write.  It's more of an on/off switch (potentially in the initialization?) than something that comes and goes.
3. There's no clear trigger or reproducer to get into this condition.  So it could be a race at the beginning or something like use of uninitialized data.

**Comment by esindril** [comment]
@bbockelm For this particular transfer that I looked into, indeed it ends up corrupted at every MB from the beginning to the end. I can ask Mihai to trigger some more transfers without a fallback to TPC push so that I can inspect the corrupted files and confirm this pattern happens all the time.

**Comment by bbockelm** [comment]
@esindril - if Mihai can reproduce it, it'd be useful to log the sequence of calls to the OFS layer for writes.  E.g.,

```
210223 08:44:45 3655051  ofs_write: 1048576@916455424 fn=/tmp/xrootd_export/foo_big_file.3
210223 08:44:45 3655051  ofs_write: 1048576@917504000 fn=/tmp/xrootd_export/foo_big_file.3
210223 08:44:45 3655051  ofs_write: 1048576@918552576 fn=/tmp/xrootd_export/foo_big_file.3
```

That'll let me know if the offset is misaligned or the size is incorrect, for example.  (And if neither is incorrect, then I can start looking at the buffering code)

**Comment by esindril** [comment]
This is transfer of a 10MB file and the corruption happens with the same pattern. Below you have the offset of the write requests as seen by the EOS OFS layer:

```
210223 16:19:12 time=1614093552.121915 func=Open                     level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=ReplicaParLayout:163           tident=? sec=      uid=0 gid=0 name= geo="" added replica_url=root://p06636710r84969.cern.ch:1095///eos/opstest/dteam/ggus150647/file.10mb_adlerf031dcb7_triumf_src_1a7d0635-873c-4aca-b7d0-b9bc360693f9?&cap.sym=v/TsLkR+AYK5TGUAdartlwDahHY=&cap.msg=EHPrLOVsWhVqqgs74McEpDkhZTiE+956HfrjZNdwoWKdd2SlQIWmIp+O7zzp409/CxrrtPOwOBIz3jwfa5sHj5ylRR2oMk+AzjWdauqGMybGKDpog+PM19wk2hVyGubnrABvXsh4WzXvukhrDieauLU7rbYNSJbaO161s1Ct499/fo6W58Yr4hxz+4McqBYXQ4r1WjZjQ9D8f3617O1Jjd4pZ8NwDN4WBKt++ueLu0UJCrd8Q1BY88rxgh1GZKnhnExxqPTbv+WRZmmbd11gHz/1qhmzKvVg1rSad1wR5Rt0C+htCwd+aFnZTiubu25hDG8DZSqx1WTQLTiRoc8/5DhMEoYAmdjvTT+ZS2z5zK8CZMIF6viSN6BEBJbaltmi3M7qvpQhCiMwhKv6AqkN6sQiWZPDguI66gUMJyOLKJv3wMxPb6GFdIl6yhNeTDuEXA4i9WBSO8M4IpF4kkOv1l3D7D0RFfhHkTBMDydwomOBpQShdCgZiyZD8d0fYyjwnqf8FU3PU4XqPcduFCvn0gqAGutwc2fFFr8FXJ9BbBlcLErkJFL8JoA5MTVDGYsOh3IwR+1szdjJGh2MZTYcAOGRihhG2mN90OcW0K83tz9GpXxiYhrCxJKHqC+OeUEM1hp6ChIJCjS+s8ZPIZZbjvBD+qhi4pmsLN6Zw0RjXdKR22Ek8gIITTUTh1z7OstyCJdylMHtc/tgAFAYqt4wyGmWXISjkCxQ/WciFUaL0owluAgBAO20xQ==&mgm.logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9&mgm.replicaindex=0&mgm.replicahead=0&mgm.etag="1278278589715841024:00000000"&mgm.id=11bd5b9ad&authz=Bearer%20MDAxNGxvY2F0aW9uIGVvc3BwcwowMDM0aWRlbnRpZmllciA4MGYzNGM4ZS05MmVkLTQ1NWItYTdmMS0xMmM4NTU4MWJhMDYKMDAxNmNpZCBuYW1lOmR0ZWFtMDAxCjAwNTJjaWQgYWN0aXZpdHk6UkVBRF9NRVRBREFUQSxVUExPQUQsRE9XTkxPQUQsREVMRVRFLE1BTkFHRSxVUERBVEVfTUVUQURBVEEsTElTVAowMDMyY2lkIGFjdGl2aXR5Ok1BTkFHRSxVUExPQUQsREVMRVRFLExJU1QsTUFOQUdFCjAwNzNjaWQgcGF0aDovZW9zL29wc3Rlc3QvZHRlYW0vZ2d1czE1MDY0Ny9maWxlLjEwbWJfYWRsZXJmMDMxZGNiN190cml1bWZfc3JjXzFhN2QwNjM1LTg3M2MtNGFjYS1iN2QwLWI5YmMzNjA2OTNmOQowMDI0Y2lkIGJlZm9yZToyMDIxLTAyLTIzVDE4OjE5OjExWgowMDJmc2lnbmF0dXJlINj5pDLBcpnRm9jrKiDHQu1hDhb7Nx8YAs70hfkjuz9ICg&mgm.path=/eos/opstest/dteam/ggus150647/file.10mb_adlerf031dcb7_triumf_src_1a7d0635-873c-4aca-b7d0-b9bc360693f9, index=0
210223 16:19:13 time=1614093553.861754 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=0 size=1048576
210223 16:19:13 time=1614093553.997379 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=1048576 size=1048576
210223 16:19:14 time=1614093554.133913 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=2097152 size=1048576
210223 16:19:14 time=1614093554.135736 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=3145728 size=1048576
210223 16:19:14 time=1614093554.137479 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=4194304 size=1048576
210223 16:19:14 time=1614093554.268718 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=5242880 size=1048576
210223 16:19:14 time=1614093554.272175 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=6291456 size=1048576
210223 16:19:14 time=1614093554.404209 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=7340032 size=1048576
210223 16:19:14 time=1614093554.405780 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=8388608 size=1048576
210223 16:19:14 time=1614093554.543100 func=write                    level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:788              tident=? sec=      uid=18118 gid=2688 name=nobody geo="" rc=1048576 offset=9437184 size=1048576
210223 16:19:14 time=1614093554.543127 func=_close                   level=INFO  logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:1149             tident=? sec=      uid=18118 gid=2688 name=nobody geo="" viaDelete=0 writeDelete=0
210223 16:19:14 time=1614093554.543139 func=VerifyChecksum           level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:2957             tident=? sec=      uid=18118 gid=2688 name=nobody geo="" isrw=1 max-offset=10485760 opensize=0
210223 16:19:14 time=1614093554.543148 func=VerifyChecksum           level=INFO  logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:3017             tident=? sec=      uid=18118 gid=2688 name=nobody geo="" (write) checksum type: adler checksum hex: 189bcc56 requested-checksum hex: -none-
210223 16:19:14 time=1614093554.543212 func=_close                   level=DEBUG logid=7b95ac2c-75ea-11eb-a4c9-fa163ec4aef9 unit=fst@p06636710r84969.cern.ch:1095 tid=00007fa7f28fc700 source=XrdFstOfsFile:1231             tident=? sec=      uid=18118 gid=2688 name=nobody geo="" checksumerror = 0, targetsizerror= 0,mMaxOffsetWritten = 10485760, targetsize = 0
```

It all looks correct to me. The corresponding FTS transfer is here:
https://fts-pilot-006.cern.ch:8449/var/log/fts3/transfers/2021-02-23/webdav.lcg.triumf.ca__eospps.cern.ch/2021-02-23-1519__webdav.lcg.triumf.ca__eospps.cern.ch__1448495912__79cbcb7e-75ea-11eb-aea2-fa163ece561c

For reference you can find attached the hexdump of the correct file and the corrupted one, if you want to look into more detail on the corruption pattern.

**Comment by esindril** [comment]
Links to the hexdump files as they are too large for github:
https://cernbox.cern.ch/index.php/s/gAWDdtNZ6FINl19
https://cernbox.cern.ch/index.php/s/hhc3059ba844q5o

**Comment by snafus** [comment]
I'm not sure if this is helpful, but below is the "extra" logging info from XrdTpcStream.cc that's able to be compiled in.
I've included just the bit around the 1MiB boundary.
The resultant file contains 6636 bytes after 1MiB that are incorrect.

`210223 17:07:52 541073 TPC_Stream::Write: Offset=1005004, Size=16384, force=0, m_offset=0
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Beginning dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Stream offset: 0
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Buffer 0: Offset=0, Size=1005004, Capacity=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Finish dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::Write: Offset=1021388, Size=16384, force=0, m_offset=0
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Beginning dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Stream offset: 0
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Buffer 0: Offset=0, Size=1021388, Capacity=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Finish dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::Write: Offset=1037772, Size=16384, force=0, m_offset=0
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Beginning dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Stream offset: 0
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Buffer 0: Offset=0, Size=1037772, Capacity=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Finish dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::Write: Offset=1054156, Size=16384, force=0, m_offset=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Beginning dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Stream offset: 1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Buffer 0: Offset=1048576, Size=5580, Capacity=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Finish dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::Write: Offset=1070540, Size=16384, force=0, m_offset=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Beginning dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Stream offset: 1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Buffer 0: Offset=1048576, Size=21964, Capacity=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Finish dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::Write: Offset=1086924, Size=16384, force=0, m_offset=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Beginning dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Stream offset: 1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Buffer 0: Offset=1048576, Size=38348, Capacity=1048576
210223 17:07:52 541073 TPC_Stream::DumpBuffers: Finish dump of stream buffers.
210223 17:07:52 541073 TPC_Stream::Write: Offset=1103308, Size=16384, force=0, m_offset=1048576
`

James

**Comment by abh3** [comment]
Hi Elvin,

Since I didn't see any changes to the actual coalescing algorithm, only 
changes to how it is structurally applied, I doubt it's fixed in 5.1. Of 
course, the code there is rather convoluted so the fix, if any, may have 
eluded me. Only a test will tell.

Andy

**Comment by bbockelm** [comment]
@snafus @esindril  - ah-ha, this is exactly what I needed!

Elvin's trace indicated that the filesystem calls were correct so the issue must be in the buffering.

James's trace indicated that the corruption was suspiciously similar to the offset size of the first partial buffer from curl.  That means the problem is in the buffer code and probably due to the partial offset.

Here's the corresponding fix: https://github.com/xrootd/xrootd/pull/1409

It turns out that we didn't offset the partial buffer, meaning that it was always "off" by however much the first curl buffer differed from exactly 16KB.

- Depending on the network speed, often curl's first read out from the socket is indeed 16KB and we're unaffected.
- While I explicitly tested this case locally, it turns out that my test files were from `/dev/zero` instead of `/dev/urandom` like I thought.

Because of the testing issue, the fact that the wrong part of the buffer was copied from was irrelevant because it happened to generate the right results!

So, can reliably reproduce now and the fix seems to work.

**Comment by snafus** [comment]
Hi @bbockelm ,
 Thanks! I merged your fix into my 5.1.0 at RAL, and from the same tests on the 10MB file, thinks look ok with dCache, EOS and DPM source sites using curl tpc. 
Cheers,
James

**Comment by bbockelm** [comment]
@esindril - the same bug appears to have been backported to 4.12.x as well (and has the same fix).

I don't want to backport it 4.12.x as I don't want to have any more 4.12.x releases -- but in case if you're desperate for something at CERN, that would be the solution.

**Comment by esindril** [comment]
@bbockelm I can also confirm that using this patch on top of 4.12.7 seems to fix the issue. I deployed this on our pre-production instance and out of 100 FTS transfers none failed. While before only 1 in 10 succeeded. I will check with Michal for the backport as both ATLAS and LHCb might be interested in this one. Thanks!

**Comment by abh3** [comment]
This has been fixed by patch  471db4e.
